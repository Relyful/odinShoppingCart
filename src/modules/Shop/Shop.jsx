import { useEffect, useState } from "react"
import Navbar from "../Navbar/Navbar"
import Footer from "../Footer/Footer"
import ShopCards from "../ShopCards/ShopCards"
import styles from "./Shop.module.css"

const useShopData = () => {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  function dataOnChange(id, increment) {
    setData((prevData) => 
      prevData.map((item) => 
        item.id === id ? {...item, cartAmmount: Math.max(0, item.cartAmmount + increment)} : item
      )
    ); 
  };

  const handleInputChange = (e, id) => {
    const newValue = e.target.value;
    let parsedValue = undefined; 
    if (newValue === "" || isNaN(newValue)) {      
      parsedValue = "";
    } else {
      parsedValue = Math.max(1, Number(newValue));
    }
    setData((prevData) => 
      prevData.map((item) => 
      item.id === id ? {...item, cartAmmount: parsedValue} : item)
    )
  }

  useEffect(() => {
    const fetchShopItems = async (retryCount = 3) => {
      try {
        const response = await fetch('https://fakestoreapi.in/api/products?limit=10', {mode: "cors"});
        if (!response.ok) {
          throw new Error('Fetch Error');
        }
        let shopData = await response.json();
        const shopDataAmm = shopData.products.map((product) => ({
          ...product, cartAmmount: 1,
        }));      

        setData(shopDataAmm);
        setError(null);
      } catch (err) {        
        setData(null);
        if (retryCount > 0) {
          fetchShopItems(retryCount - 1);
          console.log('Retrying...');
        } else {
          console.log('Failed to fetch data');
          setError(err.message);         
        }
      } finally {
        setLoading(false);             
      }
    }
    fetchShopItems();
  }, []);

  return { data, loading, error, dataOnChange, handleInputChange };
};



export default function Shop () {
  const { data, loading, error, dataOnChange, handleInputChange } = useShopData();
  const [cartItems, setCartItems] =  useState({});
  console.log(data);

  return (
    <>
      <Navbar />
      <main className={`main ${styles.main}`}>
        {loading ? <div>Loading...</div>
        : error ? <div>Error: {error}</div>
        : <ShopCards data={data} dataOnChange={dataOnChange} handleInputChange={handleInputChange} />}
      </main>
      <Footer />
    </>
  )
}