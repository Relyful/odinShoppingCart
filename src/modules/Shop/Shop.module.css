.card {
  max-width: 300px;
  max-height: 500px;
}

.main {
  display: flex;
  flex-wrap: wrap;
}

.overflow {
  overflow: hidden;
  height: 120px;
}