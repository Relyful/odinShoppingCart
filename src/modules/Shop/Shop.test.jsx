import { describe, it, expect } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { act } from 'react'
import Shop from './Shop'

describe('Shop.jsx', () => {
  it('Should render 10 inputs 1 for each item', async () => {
    act(() => {
      render(
        <MemoryRouter>
          <Shop />
        </MemoryRouter>
        );
    })

    await waitFor(() => {
      const inputs = screen.getAllByRole('spinbutton');
      expect(inputs.length).toBe(10);
    }) 
  });
})