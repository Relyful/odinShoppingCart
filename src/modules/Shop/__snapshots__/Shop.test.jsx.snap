// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop.jsx > Snapshot correct render after fetch 1`] = `
<div>
  <nav
    class="navBar"
  >
    <ul
      class="_list_2b7a49"
    >
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/shop"
        >
          Shop
        </a>
      </li>
    </ul>
    <div
      class="_cart_2b7a49"
    >
      0
      <a
        href="/"
      >
        <img
          alt="shopping cart"
          class="_cartImg_2b7a49"
          src="/src/assets/ShoppingCart.svg"
        />
      </a>
    </div>
  </nav>
  <main
    class="main _main_a55e64"
  >
    <div
      class="_card_641d47"
    >
      <h3
        class="_overflow_641d47"
      >
        Product 1
      </h3>
      <img
        class="itemPic"
      />
      <div
        class="price"
      >
        24
        â‚¬
      </div>
      <div
        class="cartRow"
      >
        <button
          type="button"
        >
          Add to Cart
        </button>
        <div
          class="ammount"
        >
          <button
            type="button"
          >
            -
          </button>
          <input
            type="number"
            value="1"
          />
          <button
            type="button"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="footer _footie_75b3b4"
  >
    <p>
      Made by Relyful for The Odin Project
    </p>
  </footer>
</div>
`;
